### DISCUSSION TITLE
Inconsistencies between RelationFilters

### DISCUSSION THREAD

**Discussion by asciant** [prisma/prisma#11923] [Q&A] [answered]
Good morning, 

I am not sure if this is a bug, or just something I am missing, but in my schema I have two seemingly identical relations, yet they both have different filters.

To give some context to the schema, it is for an online store. The store has products, products have sizes (Small, Medium, Large etc.) and the products have images (1 featured image and just general other images). The featured image (`feature`), is really just an image, so it creates a second relation between `ProductImage` and `Product`, the first relation being `images` and `ProductImage`. Ditto for `sizes` and `ProductSize` and `feature` and `ProductSize`.

Schema:
```javascript
model Product {
  id          Int            @id @default(autoincrement())
  createdAt   DateTime       @default(now())
  updatedAt   DateTime       @updatedAt
  name        String
  description String
  category    Category?      @relation(fields: [categoryId], references: [id])
  categoryId  Int?
  sizes       ProductSize[]  @relation("sizes")
  defaultSize ProductSize?   @relation("defaultSize")
  images      ProductImage[] @relation("images")
  feature     ProductImage?  @relation("feature")
  wishlist    WishList[]
  cartItem    CartItem[]
}

model ProductSize {
  id            Int                 @id @default(autoincrement())
  createdAt     DateTime            @default(now())
  updatedAt     DateTime            @updatedAt
  size          String
  stock         Int                 @default(0)
  price         Int                 @default(9999)
  inStock       Boolean             @default(false)
  product       Product?            @relation("sizes", fields: [productId], references: [id])
  productId     Int?
  defaultSize   Product?            @relation("defaultSize", fields: [defaultSizeId], references: [id])
  defaultSizeId Int?                @unique
  dimensions    PackageMeasurement?
  cartItem      CartItem[]
}

model ProductImage {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  src       String
  order     Int      @default(99)
  signature String   @unique
  product   Product? @relation("images", fields: [productId], references: [id])
  productId Int?
  feature   Product? @relation("feature", fields: [featureId], references: [id])
  featureId Int?     @unique
}
```

Prisma has generated the types based on my schema, although the filters are different, in that one allows me to use `null` as a value, although the other doesn't.

```javascript
  export type ProductSizeRelationFilter = {
    is?: ProductSizeWhereInput
    isNot?: ProductSizeWhereInput
  }

  export type ProductImageRelationFilter = {
    is?: ProductImageWhereInput | null
    isNot?: ProductImageWhereInput | null
  }
```

This means that I am not able to query products that only have a featured image and a default size. Because prisma treats `null` as a value, but `undefined` as do nothing. So the `defaultSize` filter is omitted in the query, meaning I get products that don't have a defaultsize configured.

```javascript
const products = await prisma.product.findMany({
          where: {
            AND: [
              {
                feature: { isNot: null },
                defaultSize: { isNot: undefined } // <-- I'm not allowed to use null here.
              }
            ]
          },
          select: {
            id: true,
            name: true,
            feature: {
              select: {
                src: true
              }
            }
          },
          orderBy: {
            createdAt: 'asc'
          },
          take: 8
        })
```

Is it something that I am doing wrong in my schema? Or is this a bug in the generated types?

**Comment by austincrim** [comment]
Issue created #11942

**Reply by jharrell** [reply, marked as answer]
Resolved here: https://github.com/prisma/prisma/issues/11942#issuecomment-1047506475

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
